### PR TITLE
socketpair_unix: avoid double close(), set FD_CLOEXEC

### PR DESCRIPTION
**Pull request by champtar**
We were calling the close() syscall multiple time
with the same fd number, leading to random issues like
closing containerd stream port.

In newLaunchedPlugin() we have:
```
sockets, _ := net.NewSocketPair()
defer sockets.Close()
conn, _ := sockets.LocalConn()
peerFile := sockets.PeerFile()
defer func() {
  peerFile.Close()
  if retErr != nil {
    conn.Close()
  }
}()
cmd.Start()
```

so we were doing:
```
close(local) (in LocalConn())
cmd.Start()
close(peer) (peerFile.Close())
close(local) (sockets.Close())
close(peer) (sockets.Close())
```

If the NRI plugin that we launch with cmd.Start() is not cached or
the system is a bit busy, cmd.Start() gives a large enough window
for another goroutine to open a file that will get the same fd number
as local was, thus being closed by accident.

Fix the situation by storing os.Files instead of ints, so that closing
multiple times just returns an error.

Also set FD_CLOEXEC on the sockets so we don't leak them.

Fixes 1da2cdfebfcecfb6cb0c93fc20fa3822d918f00b

### CODE REVIEW

**Comment by klihub** [comment]
> not tested with containerd / windows version also needs fixing

Yeah, we haven't really had the time to impement and test NRI for Windows, yet. That said, on Windows, SocketPair emulates a socketpair using two ends of a pre-connected TCPv4 socket, which are then stored as sys.Handle's. Shouldn't that protect us from a similar double-close error ?

**Review by klihub — APPROVED** [review]
This looks fine to me.

**Comment by champtar** [comment]
@klihub just pushed a version fixing windows and moving all the common functions in socketpair.go

File: pkg/net/socketpair_unix.go
Comment (fuweid) [inline pkg/net/socketpair_unix.go:33]: would you mind to use `unix.SOCK_CLOEXEC` here? so that we don't have to lock.

File: pkg/net/socketpair_windows.go
Comment (fuweid) [inline pkg/net/socketpair_windows.go:51]: I think the lock is too big here. would you please use a wrap to handle each syscall?

REF: https://github.com/golang/go/blob/d7df7f4fa01f1b445d835fc908c54448a63c68fb/src/net/sock_windows.go#L31

**Comment by fuweid** [comment]
Hi @champtar would you please file a ticket to track window part and focus on linux part in this pull request? I don't have too much knowledge on windows platform actually. we can ask other maintainers to take over this. Thanks

**Comment by klihub** [comment]
> Hi @champtar would you please file a ticket to track window part and focus on linux part in this pull request? I don't have too much knowledge on windows platform actually. we can ask other maintainers to take over this. Thanks

+1 I would also leave Windows out of this now and focus on fixing the bug on un*x, since we anyway don't have all the necessary functionality in place for Windows. Sorry if my earlier question/comment was misleading and could be interpreted otherwise.

File: pkg/net/socketpair_unix.go
Comment (champtar) [reply pkg/net/socketpair_unix.go:33]: unix.SOCK_CLOEXEC doesn't exists for darwin for example, this is the generic way

**Comment by champtar** [comment]
Pushed back the unix only changes, I'll open a second PR once this one is merged

File: pkg/net/socketpair_unix.go
Comment (champtar) [reply pkg/net/socketpair_unix.go:39]: I'm fine doing a linux only version, just tell me

Comment (fuweid) [reply pkg/net/socketpair_unix.go:39]: Let's focus on linux right now and add issue to track it for darwin. Thanks!

File: pkg/net/socketpair_unix.go
Comment (klihub) [inline pkg/net/socketpair_unix.go:61]: Ditto here. Isn't this also changing the semantics of PeerConn() so that it can only be called once with subsequent calls failing ? Should at least be documented.

File: pkg/net/socketpair_windows.go
Comment (champtar) [reply pkg/net/socketpair_windows.go]: Will be done in windows PR

File: pkg/net/socketpair_unix.go
Comment (champtar) [reply pkg/net/socketpair_unix.go:56]: Before this PR, a second call to LocalConn would potentially try to use a completely unrelated file. If I don't close LocalFile on of the test timeout. I'll improve documentation. I can also return nil after first *File call if preferred.

File: pkg/net/socketpair_unix.go
Comment (champtar) [reply pkg/net/socketpair_unix.go]: I've added a linux and a unix version

File: pkg/net/socketpair_unix.go
Comment (champtar) [reply pkg/net/socketpair_unix.go]: Added a small comment in the doc `// This closes LocalFile().`

**Comment by champtar** [comment]
@klihub hopefully final version

**Review by klihub — APPROVED** [review]

File: pkg/net/socketpair_windows.go
Comment (champtar) [reply pkg/net/socketpair_windows.go]: PR for windows: https://github.com/containerd/nri/pull/69

**Pull request merged**
